File: Reorder List/Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return nothing
    def reorderList(self, head):
        if head==None:
            return
        else:
            len=1
            temp=head
            while temp.next:
                len=len+1
                temp=temp.next
            self.reorder(head, len)
    def reorder(self, head, len):
        tail=head
        if len>2:
            tail=self.reorder(head.next, len-2)
        elif len==2:
            return head.next
        else:
            return head
        temp=tail.next
        tail.next=temp.next
        temp.next=head.next
        head.next=temp
        return tail
       
        
        

File: Reorder List/Reorder List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
   public void reorderList(ListNode head) {
    if (head == null) return;
    int length = 1;
    ListNode pointer = head;
    while (pointer.next != null) {
        length++;
        pointer = pointer.next;
    }
    reorder(head, length);
 }

private ListNode reorder(ListNode head, int length) {
    ListNode tail = head;
    if (length > 2) {
        tail = reorder(head.next, length - 2);
    } else if (length == 2) {
        return head.next;
    } else {
        return head;
    }

    ListNode t = tail.next;
    tail.next = t.next;
    t.next = head.next;
    head.next = t;
    return tail;   
 }
}
